File: pkg/day05/day05.rb
class Day05
  def initialize(path)
    fullpath = File.expand_path(path, __dir__)

    @rules = Hash.new { |hash, key| hash[key] = Set.new }
    @updates = []

    lines = File.readlines(fullpath).each

    loop do
      line = lines.next.chomp
      break if line.empty?

      before, after = line.split("|").map(&:to_i)
      @rules[before] << after
    end

    loop do
      line = lines.next.chomp
      break if line.empty?

      @updates << line.split(",").map(&:to_i)
    end
  end

  def part1
    sum = 0

    @updates.each do |update|
      next unless valid_update?(update)
      sum += update[update.size/2]
    end

    sum
  end

  def part2
    sum = 0

    @updates.each do |update|
      next if valid_update?(update)
      update.sort! do |a, b|
        if @rules[a].include?(b)
          -1
        elsif @rules[b].include?(a)
          1
        else
          0
        end
      end
      sum += update[update.size/2]
    end

    sum
  end

  private

  def valid_update?(update)
    preceding = Set.new
    update.each_with_index do |page, index|
      return false if @rules[page].intersect?(preceding)
      preceding << page
    end
    true
  end
end

File: pkg/day05/day05.go
package day05

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"

	"golang.org/x/exp/slices"
)

type set map[int]bool

type input struct {
	rules   map[int]set
	updates [][]int
}

func load(inputPath string) (*input, error) {
	file, err := os.Open(inputPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open input file: %w", err)
	}
	defer file.Close()

	input := input{
		rules:   map[int]set{},
		updates: [][]int{},
	}

	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		text := scanner.Text()
		if text == "" {
			break
		}

		parts := strings.Split(text, "|")

		before, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, fmt.Errorf("failed to parse before value: %w", err)
		}

		after, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("failed to parse after value: %w", err)
		}

		if _, ok := input.rules[before]; !ok {
			input.rules[before] = set{}
		}

		input.rules[before][after] = true
	}

	for scanner.Scan() {
		text := scanner.Text()
		parts := strings.Split(text, ",")

		update := make([]int, 0, len(parts))

		for _, part := range parts {
			page, err := strconv.Atoi(part)
			if err != nil {
				return nil, fmt.Errorf("failed to parse page value: %w", err)
			}

			update = append(update, page)
		}

		input.updates = append(input.updates, update)
	}

	return &input, nil
}

func (i *input) isValid(update []int) bool {
	for p, page := range update {
		// find any rules for pages after the current page
		pages, ok := i.rules[page]
		if !ok {
			continue
		}

		// check if any pages before the current page are invalid
		for _, before := range update[:p] {
			if pages[before] {
				return false
			}
		}
	}

	return true
}

func (i *input) part1() int {
	sum := 0

	for _, update := range i.updates {
		if i.isValid(update) {
			sum += update[len(update)/2]
		}
	}

	return sum
}

func (i *input) part2() int {
	sum := 0

	for _, update := range i.updates {
		if !i.isValid(update) {
			slices.SortStableFunc(update, func(a, b int) int {
				pages, ok := i.rules[a]
				if !ok {
					return 0
				}

				if pages[b] {
					return -1
				}

				return 0
			})

			sum += update[len(update)/2]
		}
	}

	return sum
}
